File: java/src/main/java/benchmarks/CalculatorAst.java
package benchmarks;

import java.util.*;

public class CalculatorAst extends Benchmark {

    interface Node {}

    static class Number implements Node {
        final long value;
        Number(long value) {
            this.value = value;
        }
    }

    static class Variable implements Node {
        final String name;
        Variable(String name) {
            this.name = name;
        }
    }

    static class BinaryOp implements Node {
        final char op;
        final Node left;
        final Node right;
        BinaryOp(char op, Node left, Node right) {
            this.op = op;
            this.left = left;
            this.right = right;
        }
    }

    static class Assignment implements Node {
        final String variable;
        final Node expr;
        Assignment(String variable, Node expr) {
            this.variable = variable;
            this.expr = expr;
        }
    }

    private long resultVal;
    private String text;
    public List<Node> expressions = new ArrayList<>();
    public long n;

    public CalculatorAst() {
        n = configVal("operations");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Calculator::Ast";
    }

    private String generateRandomProgram(long lines) {
        StringBuilder sb = new StringBuilder();
        sb.append("v0 = 1\n");

        for (int i = 0; i < 10; i++) {
            int v = i + 1;
            sb.append("v").append(v).append(" = v").append(v - 1)
            .append(" + ").append(v).append("\n");
        }

        for (long i = 0; i < lines; i++) {
            int v = (int)(i + 10);
            sb.append("v").append(v).append(" = v").append(v - 1).append(" + ");

            switch (Helper.nextInt(10)) {
            case 0:
                sb.append("(v").append(v - 1).append(" / 3) * 4 - ").append(i)
                .append(" / (3 + (18 - v").append(v - 2).append(")) % v")
                .append(v - 3).append(" + 2 * ((9 - v").append(v - 6)
                .append(") * (v").append(v - 5).append(" + 7))");
                break;
            case 1:
                sb.append("v").append(v - 1).append(" + (v").append(v - 2)
                .append(" + v").append(v - 3).append(") * v").append(v - 4)
                .append(" - (v").append(v - 5).append(" /  v").append(v - 6).append(")");
                break;
            case 2:
                sb.append("(3789 - (((v").append(v - 7).append(")))) + 1");
                break;
            case 3:
                sb.append("4/2 * (1-3) + v").append(v - 9).append("/v").append(v - 5);
                break;
            case 4:
                sb.append("1+2+3+4+5+6+v").append(v - 1);
                break;
            case 5:
                sb.append("(99999 / v").append(v - 3).append(")");
                break;
            case 6:
                sb.append("0 + 0 - v").append(v - 8);
                break;
            case 7:
                sb.append("((((((((((v").append(v - 6).append(")))))))))) * 2");
                break;
            case 8:
                sb.append(i).append(" * (v").append(v - 1).append("%6)%7");
                break;
            case 9:
                sb.append("(1)/(0-v").append(v - 5).append(") + (v").append(v - 7).append(")");
                break;
            }
            sb.append("\n");
        }

        return sb.toString();
    }

    @Override
    public void prepare() {
        text = generateRandomProgram(n);
    }

    static class Parser {

        private static final char CHAR_EOF = '\0';
        private static final char CHAR_PLUS = '+';
        private static final char CHAR_MINUS = '-';
        private static final char CHAR_STAR = '*';
        private static final char CHAR_SLASH = '/';
        private static final char CHAR_PERCENT = '%';
        private static final char CHAR_LPAREN = '(';
        private static final char CHAR_RPAREN = ')';
        private static final char CHAR_EQUALS = '=';
        private static final char CHAR_ZERO = '0';
        private static final char CHAR_NINE = '9';
        private static final char CHAR_A_LOWER = 'a';
        private static final char CHAR_Z_LOWER = 'z';
        private static final char CHAR_A_UPPER = 'A';
        private static final char CHAR_Z_UPPER = 'Z';
        private static final char CHAR_SPACE = ' ';
        private static final char CHAR_TAB = '\t';
        private static final char CHAR_NEWLINE = '\n';
        private static final char CHAR_CR = '\r';

        private final String input;
        private int pos;
        private char currentChar;
        final List<Node> expressions = new ArrayList<>();

        Parser(String input) {
            this.input = input;
            this.pos = 0;
            this.currentChar = input.length() > 0 ? input.charAt(0) : CHAR_EOF;
        }

        void parse() {
            while (currentChar != CHAR_EOF) {
                skipWhitespace();
                if (currentChar == CHAR_EOF) break;

                expressions.add(parseExpression());

                skipWhitespace();
                while (currentChar == CHAR_NEWLINE) {
                    advance();
                    skipWhitespace();
                }
            }
        }

        private Node parseExpression() {
            Node node = parseTerm();

            while (true) {
                skipWhitespace();

                if (currentChar == CHAR_PLUS || currentChar == CHAR_MINUS) {
                    char op = currentChar;
                    advance();
                    Node right = parseTerm();
                    node = new BinaryOp(op, node, right);
                } else {
                    break;
                }
            }

            return node;
        }

        private Node parseTerm() {
            Node node = parseFactor();

            while (true) {
                skipWhitespace();

                if (currentChar == CHAR_STAR || currentChar == CHAR_SLASH || currentChar == CHAR_PERCENT) {
                    char op = currentChar;
                    advance();
                    Node right = parseFactor();
                    node = new BinaryOp(op, node, right);
                } else {
                    break;
                }
            }

            return node;
        }

        private Node parseFactor() {
            skipWhitespace();

            if (isDigit(currentChar)) {
                return parseNumber();
            } else if (isLetter(currentChar)) {
                return parseVariable();
            } else if (currentChar == CHAR_LPAREN) {
                advance();
                Node node = parseExpression();
                skipWhitespace();
                if (currentChar == CHAR_RPAREN) {
                    advance();
                }
                return node;
            } else {
                advance();
                return new Number(0);
            }
        }

        private Node parseNumber() {
            long value = 0;
            while (isDigit(currentChar)) {
                value = value * 10 + (currentChar - CHAR_ZERO);
                advance();
            }
            return new Number(value);
        }

        private Node parseVariable() {
            int start = pos;
            while (isLetter(currentChar) || isDigit(currentChar)) {
                advance();
            }
            String varName = input.substring(start, pos);

            skipWhitespace();
            if (currentChar == CHAR_EQUALS) {
                advance();
                Node expr = parseExpression();
                return new Assignment(varName, expr);
            }

            return new Variable(varName);
        }

        private void advance() {
            pos++;
            if (pos >= input.length()) {
                currentChar = CHAR_EOF;
            } else {
                currentChar = input.charAt(pos);
            }
        }

        private void skipWhitespace() {
            while (isWhitespace(currentChar)) {
                advance();
            }
        }

        private boolean isDigit(char c) {
            return c >= CHAR_ZERO && c <= CHAR_NINE;
        }

        private boolean isLetter(char c) {
            return (c >= CHAR_A_LOWER && c <= CHAR_Z_LOWER) ||
                   (c >= CHAR_A_UPPER && c <= CHAR_Z_UPPER);
        }

        private boolean isWhitespace(char c) {
            return c == CHAR_SPACE || c == CHAR_TAB || c == CHAR_NEWLINE || c == CHAR_CR;
        }
    }

    @Override
    public void run(int iterationId) {
        Parser parser = new Parser(text);
        parser.parse();
        expressions = parser.expressions;
        resultVal += expressions.size();

        if (!expressions.isEmpty() && expressions.get(expressions.size() - 1) instanceof Assignment) {
            Assignment assign = (Assignment) expressions.get(expressions.size() - 1);
            resultVal += Helper.checksum(assign.variable);
        }
    }

    @Override
    public long checksum() {
        return resultVal;
    }

    public List<Node> getExpressions() {
        return expressions;
    }
}

File: kotlin/src/main/kotlin/benchmarks/CalculatorAst.kt
package benchmarks

import Benchmark

class CalculatorAst : Benchmark() {
    sealed interface Node

    data class Number(
        val value: Long,
    ) : Node

    data class Variable(
        val name: String,
    ) : Node

    data class BinaryOp(
        val op: Char,
        val left: Node,
        val right: Node,
    ) : Node

    data class Assignment(
        val variable: String,
        val expr: Node,
    ) : Node

    var n = configVal("operations")
    private var resultVal: UInt = 0u
    private lateinit var text: String
    lateinit var expressions: List<Node>

    private fun generateRandomProgram(lines: Long = 1000): String =
        buildString {
            append("v0 = 1\n")
            for (i in 0 until 10) {
                val v = i + 1
                append("v$v = v${v - 1} + $v\n")
            }
            for (i in 0 until lines) {
                val v = i + 10
                append("v$v = v${v - 1} + ")
                when (Helper.nextInt(10)) {
                    0 -> append("(v${v - 1} / 3) * 4 - $i / (3 + (18 - v${v - 2})) % v${v - 3} + 2 * ((9 - v${v - 6}) * (v${v - 5} + 7))")
                    1 -> append("v${v - 1} + (v${v - 2} + v${v - 3}) * v${v - 4} - (v${v - 5} /  v${v - 6})")
                    2 -> append("(3789 - (((v${v - 7})))) + 1")
                    3 -> append("4/2 * (1-3) + v${v - 9}/v${v - 5}")
                    4 -> append("1+2+3+4+5+6+v${v - 1}")
                    5 -> append("(99999 / v${v - 3})")
                    6 -> append("0 + 0 - v${v - 8}")
                    7 -> append("((((((((((v${v - 6})))))))))) * 2")
                    8 -> append("$i * (v${v - 1}%6)%7")
                    9 -> append("(1)/(0-v${v - 5}) + (v${v - 7})")
                }
                append("\n")
            }
        }

    override fun prepare() {
        text = generateRandomProgram(n)
    }

    private class Parser(
        private val input: String,
    ) {
        companion object {
            private const val CHAR_EOF = '\u0000'
            private const val CHAR_PLUS = '+'
            private const val CHAR_MINUS = '-'
            private const val CHAR_STAR = '*'
            private const val CHAR_SLASH = '/'
            private const val CHAR_PERCENT = '%'
            private const val CHAR_LPAREN = '('
            private const val CHAR_RPAREN = ')'
            private const val CHAR_EQUALS = '='
            private const val CHAR_ZERO = '0'
            private const val CHAR_NINE = '9'
            private const val CHAR_A_LOWER = 'a'
            private const val CHAR_Z_LOWER = 'z'
            private const val CHAR_A_UPPER = 'A'
            private const val CHAR_Z_UPPER = 'Z'
            private const val CHAR_SPACE = ' '
            private const val CHAR_TAB = '\t'
            private const val CHAR_NEWLINE = '\n'
            private const val CHAR_CR = '\r'
        }

        private var pos = 0
        private var currentChar = if (input.isNotEmpty()) input[0] else CHAR_EOF
        val expressions = mutableListOf<Node>()

        fun parse(): List<Node> {
            while (currentChar != CHAR_EOF) {
                skipWhitespace()
                if (currentChar == CHAR_EOF) break

                expressions.add(parseExpression())

                skipWhitespace()
                while (currentChar == CHAR_NEWLINE) {
                    advance()
                    skipWhitespace()
                }
            }
            return expressions
        }

        private fun parseExpression(): Node {
            var node = parseTerm()

            while (true) {
                skipWhitespace()

                if (currentChar == CHAR_PLUS || currentChar == CHAR_MINUS) {
                    val op = currentChar
                    advance()
                    val right = parseTerm()
                    node = BinaryOp(op, node, right)
                } else {
                    break
                }
            }

            return node
        }

        private fun parseTerm(): Node {
            var node = parseFactor()

            while (true) {
                skipWhitespace()

                if (currentChar == CHAR_STAR || currentChar == CHAR_SLASH || currentChar == CHAR_PERCENT) {
                    val op = currentChar
                    advance()
                    val right = parseFactor()
                    node = BinaryOp(op, node, right)
                } else {
                    break
                }
            }

            return node
        }

        private fun parseFactor(): Node {
            skipWhitespace()

            return when {
                isDigit(currentChar) -> {
                    parseNumber()
                }

                isLetter(currentChar) -> {
                    parseVariable()
                }

                currentChar == CHAR_LPAREN -> {
                    advance()
                    val node = parseExpression()
                    skipWhitespace()
                    if (currentChar == CHAR_RPAREN) {
                        advance()
                    }
                    node
                }

                else -> {
                    advance()
                    Number(0)
                }
            }
        }

        private fun parseNumber(): Node {
            var value = 0L
            while (isDigit(currentChar)) {
                value = value * 10 + (currentChar - CHAR_ZERO)
                advance()
            }
            return Number(value)
        }

        private fun parseVariable(): Node {
            val start = pos
            while (isLetter(currentChar) || isDigit(currentChar)) {
                advance()
            }
            val varName = input.substring(start, pos)

            skipWhitespace()
            if (currentChar == CHAR_EQUALS) {
                advance()
                val expr = parseExpression()
                return Assignment(varName, expr)
            }

            return Variable(varName)
        }

        private fun advance() {
            pos++
            if (pos >= input.length) {
                currentChar = CHAR_EOF
            } else {
                currentChar = input[pos]
            }
        }

        private fun skipWhitespace() {
            while (isWhitespace(currentChar)) {
                advance()
            }
        }

        private fun isDigit(char: Char): Boolean = char >= CHAR_ZERO && char <= CHAR_NINE

        private fun isLetter(char: Char): Boolean =
            (char >= CHAR_A_LOWER && char <= CHAR_Z_LOWER) ||
                (char >= CHAR_A_UPPER && char <= CHAR_Z_UPPER)

        private fun isWhitespace(char: Char): Boolean = char == CHAR_SPACE || char == CHAR_TAB || char == CHAR_NEWLINE || char == CHAR_CR
    }

    override fun run(iterationId: Int) {
        val parser = Parser(text)
        expressions = parser.parse()
        resultVal += expressions.size.toUInt()
        val last = expressions.lastOrNull()
        if (last is Assignment) {
            resultVal += Helper.checksum(last.variable)
        }
    }

    override fun checksum(): UInt = resultVal

    override fun name(): String = "Calculator::Ast"
}
